### ISSUE TITLE
Running into SSL memory leaks

### ISSUE THREAD

**Issue by HappyCerberus** [libgit2/libgit2#5505]
Both ASAN and valgrind complain about memory leaks coming from libgit use of SSL.
I'm running libgit version 1.0.0.

```
==16329== 61,206 (1,488 direct, 59,718 indirect) bytes in 6 blocks are definitely lost in loss record 238 of 238
==16329==    at 0x483A7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==16329==    by 0x4E87D98: OPENSSL_malloc (mem.c:108)
==16329==    by 0x4E2EC09: asn1_item_ex_combine_new (tasn_new.c:182)
==16329==    by 0x4E2EC8F: ASN1_item_new (tasn_new.c:79)
==16329==    by 0x4EAB5E6: X509_parse_from_buffer (x_x509.c:160)
==16329==    by 0x4DB3E54: bssl::ssl_crypto_x509_session_cache_objects(ssl_session_st*) (ssl_x509.cc:304)
==16329==    by 0x4DBE3EA: bssl::tls13_process_certificate(bssl::SSL_HANDSHAKE*, bssl::SSLMessage const&, bool) (tls13_both.cc:316)
==16329==    by 0x4DC04F2: do_read_server_certificate (tls13_client.cc:556)
==16329==    by 0x4DC04F2: bssl::tls13_client_handshake(bssl::SSL_HANDSHAKE*) (tls13_client.cc:778)
==16329==    by 0x4D96207: do_tls13 (handshake_client.cc:794)
==16329==    by 0x4D96207: bssl::ssl_client_handshake(bssl::SSL_HANDSHAKE*) (handshake_client.cc:1759)
==16329==    by 0x4D94625: bssl::ssl_run_handshake(bssl::SSL_HANDSHAKE*, bool*) (handshake.cc:676)
==16329==    by 0x4DAC600: SSL_do_handshake (ssl_lib.cc:890)
==16329==    by 0x4D115CB: openssl_connect (openssl.c:647)
```

### Reproduction steps

I'm running v1.0.0 with BoringSSL if that matters.

### Expected behavior

no leaks

### Actual behavior

it's leaking

### Version of libgit2 (release number or SHA1)

v1.0.0

### Operating system(s) tested

Ubuntu

**Comment by ethomson** [comment]
Are you sure that this memory leak results from our improper use of the APIs?

**Comment by pks-t** [comment]
Yeah, this is one of those issues I've repeatedly debugged and tried to fix with no results whatsoever, which led me to the conclusion that this really is a memory leak in openssl. I know you're not using openssl but boringssl, but it is a fork of openssl and thus the likelihood is high that it's got some of the same leaks.

Anyway, I'm still not sure about whom to blame. Are you able to provide a small reproducer that shows the problems? That'd make it a lot easier to make up our mind.

**Comment by HappyCerberus** [comment]
This is the code snippet that reproduces this problem reliably, can you try it with your local setups, so I can rule out some local weirdness in my setup?

```
#include "include/git2/clone.h"
#include "include/git2/errors.h"
#include "include/git2/global.h"
#include "include/git2/types.h"

int libgit2_decide_on_failing_cert_check(git_cert*, int, const char*, void*) {
    return GIT_PASSTHROUGH;
}

int main() {
  git_libgit2_init();
  git_clone_options clone_options = GIT_CLONE_OPTIONS_INIT;
  clone_options.fetch_opts.callbacks.certificate_check = libgit2_decide_on_failing_cert_check;

  git_repository* repo = nullptr;
  int ret = git_clone(&repo, "https://github.com/HappyCerberus/libgit2_problems", "/tmp/test_xxx", &clone_options);
  git_repository_free(repo);

  git_libgit2_shutdown();
  return ret;
}
```

**Comment by ethomson** [comment]
Thanks for the minimum reproducer.  Super helpful.  I suspect that this is a libgit2 issue.

**Comment by pks-t** [comment]
On Fri, May 15, 2020 at 07:40:20AM -0700, Edward Thomson wrote:
> Thanks for the minimum reproducer.  Super helpful.  I suspect that this is a libgit2 issue.

Indeed, it was mighty helpful! And yes, it's a libgit2 issue which is
being fixed by #5522, so please have a look.
